core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+
+								if newURL, ok := toSubdomainURL(hostname, pathPrefix+r.URL.Path, r.URL.RawQuery); ok {
+									// Redirect to CID fixed inside of toSubdomainURL()
+									http.Redirect(w, r, newURL, http.StatusMovedPermanently)
+									return
+								}
+							}
+						}
+
+						// Rewrite the path to not use subdomains
+						r.URL.Path = pathPrefix + r.URL.Path
+						// Serve path request
+						childMux.ServeHTTP(w, r)
+						return
+					}
+					// If not, resource does not exist on the subdomain gateway, return 404
+					http.NotFound(w, r)
+					return
+				}
+			}
+			// We don't have a known gateway. Fallback on DNSLink lookup
+
+			// Wildcard HTTP Host check:
+			// 1. is wildcard DNSLink enabled (Gateway.NoDNSLink=false)?
+			// 2. does Host header include a fully qualified domain name (FQDN)?
+			// 3. does DNSLink record exist in DNS?
+			if !cfg.Gateway.NoDNSLink && isDNSLinkRequest(r, n) {
+				// rewrite path and handle as DNSLink
+				r.URL.Path = "/ipns/" + stripPort(r.Host) + r.URL.Path
+				childMux.ServeHTTP(w, r)
+				return
+			}
+
+			// else, treat it as an old school gateway, I guess.
+			childMux.ServeHTTP(w, r)
+		})
+		return childMux, nil
+	}
+}
+
+// isDNSLinkRequest returns bool that indicates if request
+// should return data from content path listed in DNSLink record (if exists)
+func isDNSLinkRequest(r *http.Request, n *core.IpfsNode) bool {
+	ctx, cancel := context.WithCancel(n.Context())
+	defer cancel()
+
+	fqdn := stripPort(r.Host)
+	if len(fqdn) > 0 && isd.IsDomain(fqdn) {
+		// Confirm DNSLink was not disabled for the fqdn or globally
+		name := "/ipns/" + fqdn
+		// check if DNSLink exists
+		_, err := n.Namesys.Resolve(ctx, name, nsopts.Depth(1))
+		if err == nil || err == namesys.ErrResolveRecursion {
+			return true
+		}
+	}
+	return false
+}
+
+func isSubdomainNamespace(ns string) bool {
+	switch ns {
+	case "ipfs", "ipns", "p2p", "ipld", "api":
+		return true
+	default:
+		return false
+	}
+}
+
+// Parses Host header of a subdomain-based URL and returns it's components
+// Note: hostname is host + optional port
+func parseSubdomains(hostHeader string) (hostname, ns, rootID string, ok bool) {
+	parts := subdomainGatewayRegex.FindStringSubmatch(hostHeader)
+	if len(parts) < 4 || !isSubdomainNamespace(parts[2]) {
+		return "", "", "", false
+	}
+	return parts[3], parts[2], parts[1], true
+}
+
+// Converts a hostname/path to a subdomain-based URL, if applicable.
+func toSubdomainURL(hostname, path string, query string) (url string, ok bool) {
+	parts := strings.SplitN(path, "/", 4)
+
+	var ns, rootID, rest string
+	switch len(parts) {
+	case 4:
+		rest = parts[3]
+		fallthrough
+	case 3:
+		ns = parts[1]
+		rootID = parts[2]
+	default:
+		return "", false
+	}
+
+	if !isSubdomainNamespace(ns) {
+		return "", false
+	}
+
+	// add prefix if query is present
+	if query != "" {
+		query = "?" + query
+	}
+
+	if ns == "api" || ns == "p2p" {
+		// API and P2P proxy use the same paths on subdomains:
```

Comment:
I'd actually expect P2P to use `https://http.peerid.p2p.domain/$FORWARDED_PATH`.

Reply:
Or just `https://peerid.p2p.domain/http/$FORWARDED_PATH`. Honestly, that's probably simpler to parse. The downside is that base for `$FORWARDED_PATH` will change (like it currently does) but that's not the end of the world.

Reply:
There are some practical contraints we need to consider:

Wildcard TLS certificate will work only for a single subdomain level of "unknowns" (`*.p2p.domain`) so if we want people to be able to set up TLS for P2P proxy at a public gateway, then `https://peerid.p2p.domain/` is the max depth we can go.

Will look into this and comment here later.

Reply:
WFM. We could also do `*.http.p2p.domain`.

Reply:
For the sake of simplicity I made `/p2p/` work the same as `/ipfs/`:

- `/p2p/$peerid/http/index.txt` is represented as `$peerid.p2p.example.com/http/index.txt`
- `/p2p/$peerid/x/custom/http/index.txt` is represented as `$peerid.p2p.example.com/x/custom/http/index.txt`
- `example.com/p2p/$peerid/http/index.txt` redirects to  `$peerid.p2p.example.com/http/index.txt` (and gets same libp2p-key fix as `/ipns/`)

See tests in 835a8f06f977c7f4feb21dfa75c2668772617e05

Reply:
Something that bothers me is that `/http/` is pretty meaningless in this context.
What would make `/p2p` super useful is to **make `/http/` implicit in subdomains**.

That way content path would match URIs `/` and relative paths would just work (`/index.txt`). This would make this feature great for showing local web app to someone while behind NAT. 

To illustrate, would it be ok to make:

-  `/p2p/$peerid/http/index.txt` into `$peerid.p2p.example.com/index.txt` 
-  `/p2p/$peerid/x/custom/http/index.txt` into `$peerid.p2p.example.com/x/custom/index.txt` 

?

Reply:
> Something that bothers me is that /http/ is pretty meaningless/redundant in HTTP subdomain gateway context

I believe the main concerns were:

* A websocket socket proxy. This is basically "libp2p daemon in a gateway".
  1. Open a websocket to `http://localhost/p2p/Qm.../ws/some/protocol/name`
  2. Socket gets forwarded to a service running at `/some/protocol/name`.
* Custom protocols.

Wouldn't your proposal be ambiguous?

---

Note: this is a very low-priority decision. I'm fine with https://github.com/ipfs/go-ipfs/pull/6096#discussion_r392249872 then breaking it later.

Reply:
Yes, let's keep this PR's surface small and leave `/p2p/` handling as described in https://github.com/ipfs/go-ipfs/pull/6096#discussion_r392249872 for now.

In the future we could improve it without ambiguouity by using different `ns` subdomain for `/p2p/$peerid/http` proxy (eg. `$peerid.p2p-http.example.com/index.txt`)